javatests/arcs/core/analysis/ClaimDeductionTest.kt

```diff
@@ -0,0 +1,211 @@
+        singletons = mapOf(
+          "a" to FieldType.InlineEntity("BarHash"),
+          "b" to FieldType.Text,
+          "c" to FieldType.Number
+        ),
+        collections = emptyMap()
+      ),
+      hash = "FooHousePetHash"
+    ))),
+    expression = PaxelParser.parse(
+      """
+      new Foo {
+        a: new Bar {
+          x: input.cat, 
+          y: input.dog
+        },
+        b: input.foo,
+        c: 5
+      }
+      """.trimIndent()
+    )
+  )
+
+  @Test
+  fun particle_with_no_expressions_produces_no_claims() {
+    val connections = mapOf(
+      "input" to inputPetConnection,
+      "output" to outputPetConnection.copy(expression = null)
+    )
+    val particleSpec = ParticleSpec(
+      name = "FooHousePets",
+      location = "arcs.analysis.FooHousePets",
+      connections = connections
+    )
+    val particle = Recipe.Particle(particleSpec, emptyList())
+    val actual = RecipeGraph.Node.Particle(particle).deduceClaims()
+
+    assertThat(actual).isEmpty()
+  }
+
+  @Test
+  fun expressions_must_use_new_object_as_root() {
+    val connections = mapOf(
+      "input" to inputPetConnection,
+      "output" to outputPetConnection.copy(expression = PaxelParser.parse(
+        "input.cat + input.dog"
+      ))
+    )
+    val particleSpec = ParticleSpec(
+      name = "FooHousePets",
+      location = "arcs.analysis.FooHousePets",
+      connections = connections
+    )
+    val particle = Recipe.Particle(particleSpec, emptyList())
+
+    val e = assertThrows(IllegalArgumentException::class) {
+      RecipeGraph.Node.Particle(particle).deduceClaims()
+    }
+    assertThat(e).hasMessageThat().contains("Expression on 'FooHousePets.output' is invalid.")
+  }
+
+  @Test
+  fun deduction_fails_when_expression_references_a_missing_handle() {
+    val connections = mapOf(
+      "input" to inputPetConnection,
+      "output" to outputPetConnection.copy(expression = PaxelParser.parse(
+        "new Foo {a: notAHandle.cat}"
+      ))
+    )
+    val particleSpec = ParticleSpec(
+      name = "FooHousePets",
+      location = "arcs.analysis.FooHousePets",
+      connections = connections
+    )
+    val particle = Recipe.Particle(particleSpec, emptyList())
+
+    val e = assertThrows(IllegalArgumentException::class) {
+      RecipeGraph.Node.Particle(particle).deduceClaims()
+    }
+    assertThat(e).hasMessageThat().contains(
+      "Particle 'FooHousePets' does not have a handle connection called 'notAHandle'."
+    )
+  }
+
+  @Test
+  fun particle_creates_claims_from_inputs() {
+    val connections = mapOf(
+      "input" to inputPetConnection,
+      "output" to outputPetConnection
+    )
+    val particleSpec = ParticleSpec(
+      name = "FooHousePets",
+      location = "arcs.analysis.FooHousePets",
+      connections = connections
+    )
+    val particle = Recipe.Particle(particleSpec, emptyList())
+    val actual = RecipeGraph.Node.Particle(particle).deduceClaims()
+
+    assertThat(actual).isEqualTo(
```

Comment:
This can make the tests flaky as the order should not matter. Use `containsExactly` instead. Make the change in other places as well.